### ISSUE TITLE
A mid-epoch resume re-shuffles the epoch: the DataLoader skip drops the wrong samples (measured 15 of 20 wrong)

### ISSUE THREAD

**Issue by wyim-pgl** [wyim-pgl/transgenic#59]
Found while working the #53 checkbox "check whether the final incomplete accumulation group is dropped without an optimizer step". The accumulation answer is benign; what the same read turned up is not.

## A mid-epoch resume re-shuffles the epoch, and the skip then drops the wrong samples

`train_HyenaTransgenic.py:163` builds `Accelerator(mixed_precision="bf16")` with no `DataLoaderConfiguration`, so `use_seedable_sampler` is off. The training loader is `shuffle=True` (`:177`) with no generator, seeded once by `torch.manual_seed(seed)` at `:172`. On resume, `accelerator.load_state` restores the RNG *as it was mid-epoch* (`:281`), and the loop then skips by DataLoader index:

    if epoch == start_epoch and step < resume_step:
        continue                                    # :357-358

The permutation is drawn when the iterator is created, from the global RNG. At the original epoch's start that RNG held one state; on resume it holds the mid-epoch state that was checkpointed. Different state, different permutation — and `resume_step` then skips the first N entries of a *different* order.

Measured on pgl-gpu (torch 2.5.1, accelerate 1.14.0), 64 samples, resume at 20:

    plain DataLoader                            same order: False   15 seen twice, 15 never seen
    accelerate prepare (default, current code)  same order: False   15 seen twice, 15 never seen
    accelerate prepare (use_seedable_sampler)   same order: True     0 seen twice,  0 never seen

Three quarters of the skipped region is wrong. Script: `~/resume_shuffle.py` on pgl-gpu.

At B5 scale the train split is 198,171 tiles at batch 1, and `--save-every-n-steps` defaults to 200 optimizer steps = 19,200 micro-batches between `latest_state` saves. So a preemption resume mid-epoch re-sees roughly half of the already-processed portion of that epoch and never sees an equal number. This is the normal path, not an edge case: A28 has the job chain auto-resubmitting, and ACCESS preemption is what `latest_state` exists for. Nothing reports it — epoch loss, validation loss and the checkpoint rule all stay perfectly plausible.

The fix is one argument:

    from accelerate.utils import DataLoaderConfiguration
    accelerator = Accelerator(mixed_precision="bf16",
                              dataloader_config=DataLoaderConfiguration(use_seedable_sampler=True))

which makes Accelerate use its `SeedableRandomSampler`, whose permutation depends on the seed and epoch rather than on the ambient RNG. Verified above: 0 duplicated, 0 missed. Accelerate's own `skip_first_batches` would be the more idiomatic replacement for the hand-rolled skip and is worth folding in at the same time, but the sampler is the part that changes the data.

This should land before the first B5 seed rather than after, since it changes what each epoch actually trains on after any resume.

## The accumulation checkbox itself

`micro_done` (`:348`) is initialised once before the epoch loop and never reset, and the optimizer steps on `micro_done % accumulation_steps == 0` (`:392`). So an incomplete group at an epoch boundary is not dropped — its gradients are carried into the first optimizer step of the next epoch (evaluation runs under `no_grad` and does not disturb them). Only the residual at the very end of the final epoch is discarded, which is at most 95 micro-batches of 4.36 M.

One consequence worth recording: `steps_per_epoch = ceil(len(train_dl) / accumulation_steps)` (`:246`) assumes each epoch starts a fresh group, so `t_total` over-counts. With 198,171 micro-batches, accumulation 96 and 22 epochs, `t_total` is 2,065 x 22 = 45,430 while the run actually takes floor(198,171 x 22 / 96) = 45,414 optimizer steps. The linear decay therefore ends at about 0.035 % of peak instead of 0, and warmup is 5 % of a slightly wrong total. Negligible, but it is a discrepancy between the schedule the recipe describes and the one that runs, and it is free to fix once the counter semantics are settled.

Note also that on resume `micro_done` restarts at 0 while `global_step` is restored from `meta.json`, so the first optimizer step after a resume comes a full `accumulation_steps` micro-batches in regardless of where the previous process left off inside its group. That is the right behaviour — the partial group's gradients died with the process — but it means the accumulation-group boundaries shift at every resume, which is worth knowing when reading a loss curve across a job chain.
